### PR TITLE
[FLINK-21380][coordination] Hide terminal ExecutionGraph in StateWithExecutionGraph

### PR DESCRIPTION
**Pull request by zentol**
Hides the fact that the ExecutionGraph can reach a globally-terminal state while the DeclarativeScheduler is in a Restarting/Canceling/Failing/Executing state.
This can happen because the transition into a globally-terminal state in the EG and the scheduler transition to WaitingForResources/Finished does not happen atomically (mainthread-wise).

This should not happen during Restarting because it would break the contract that a globally-terminal job never transitions into another state.
As for Restarting/Canceling/Failing/Executing, this is mostly for consistency; we should ensure that the scheduler has a chance to cleanup whatever it wants before we communicate to the outside that the job is done.

In short, `ArchivedExecutionGraph#createFrom` now accepts an optional non-terminal JobStatus that overrides the state of the ArchivedExecutionGraph. Additionally, if set, the timestamps for all globally-terminal state transitions are removed (or rather not copied).
All `StateWithExecutionGraph` classes now also pin the job state within `getJobStatus()`.

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ArchivedExecutionGraph.java [hunk 1]
```diff
@@ -335,9 +352,16 @@ public static ArchivedExecutionGraph createFrom(ExecutionGraph executionGraph) {
 
         final long[] timestamps = new long[JobStatus.values().length];
 
+        // if the state is overridden with a non-globally-terminal state then we need to erase
+        // traces of globally-terminal states
+        final boolean clearGloballyTerminalStateTimestamps =
+                statusOverride != null && !statusOverride.isGloballyTerminalState();
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ArchivedExecutionGraph.java:358]: I think we check in the beginning of this method whether `statusOverride` is not a globally terminal state if it is not `null`.

**Review by tillrohrmann — APPROVED** [review]
Thanks for creating this PR @zentol. I think your solution is a good compromise between being minimal invasive and solving the problem. In the long run when we replace the `EG` implementation in the `DeclarativeScheduler` we might be able to solve it differently. +1 for merging.
